src/Material/Tooltip.elm

```diff
@@ -0,0 +1,523 @@
+  , offsetWidth : Float
+  , offsetHeight : Float
+  }
+
+
+{-| Default DOMState constructor
+-}
+defaultDOMState : DOMState
+defaultDOMState =
+  { rect = { left = 0, top = 0, width = 0, height = 0 }
+  , offsetWidth = 0
+  , offsetHeight = 0
+  }
+
+
+{-| Calculates the position of the tooltip based on the event
+and the requested position
+-}
+calculatePos : Position -> DOMState -> Pos
+calculatePos pos domState =
+  let
+    -- _ = Debug.log "Calculating position for " domState
+    props =
+      domState.rect
+
+    offsetWidth =
+      domState.offsetWidth
+
+    offsetHeight =
+      domState.offsetHeight
+
+    left =
+      props.left + (props.width / 2)
+
+    top =
+      props.top + (props.height / 2)
+
+    marginLeft =
+      -1 * (offsetWidth / 2)
+
+    marginTop =
+      -1 * (offsetHeight / 2)
+
+    out =
+      case pos of
+        Left ->
+          { left = props.left - offsetWidth - 10
+          , top =
+              if (top + marginTop < 0) then
+                0
+              else
+                top
+          , marginTop =
+              if (top + marginTop < 0) then
+                0
+              else
+                marginTop
+          , marginLeft = 0
+          }
+
+        Right ->
+          { left = props.left + props.width + 10
+          , top =
+              if (top + marginTop < 0) then
+                0
+              else
+                top
+          , marginTop =
+              if (top + marginTop < 0) then
+                0
+              else
+                marginTop
+          , marginLeft = 0
+          }
+
+        Top ->
+          { left =
+              if (left + marginLeft < 0) then
+                0
+              else
+                left
+          , top = props.top - offsetHeight - 10
+          , marginTop = 0
+          , marginLeft =
+              if (left + marginLeft < 0) then
+                0
+              else
+                marginLeft
+          }
+
+        Bottom ->
+          { left =
+              if (left + marginLeft < 0) then
+                0
+              else
+                left
+          , top = props.top + props.height + 10
+          , marginTop = 0
+          , marginLeft =
+              if (left + marginLeft < 0) then
+                0
+              else
+                marginLeft
+          }
+  in
+    out
+
+
+{-| Component update.
+-}
+update : Msg -> Model -> ( Model, Cmd Msg )
+update action model =
+  case action of
+    Enter dom ->
+      ( { model | isActive = True, domState = dom }, none )
+
+    Leave ->
+      ( { model | isActive = False }, none )
+
+
+{-| Tries and get the next sibling that is available and use the given decoder on it
+-}
+sibling : Json.Decoder a -> Json.Decoder a
+sibling d =
+  let
+    createPath depth =
+      let
+        parents =
+          List.repeat depth "parentElement"
+      in
+        ([ "target" ] ++ parents ++ [ "nextSibling" ])
+
+    paths =
+      List.map createPath [0..4]

```

Comment:
You look for the tooltip element among the next 5 siblings, right? Is there are reason we don't just demand that it's the following element?